tests/unit_tests/test_agent_0/test_agent_0.py

```diff
@@ -165,14 +168,71 @@ def check_agent_task(agent_task, *args, **kwargs):
         with self.assertRaises(ValueError):
             agent_0._start_sub_agents()
 
-# ------------------------------------------------------------------------------
 
+    # --------------------------------------------------------------------------
+    #
+    @mock.patch.object(Agent_0, '__init__', return_value=None)
+    def test_start_services(self, mocked_init):
+
+        advanced_services = list()
+
+        def local_advance(services, publish, push):
+
+            nonlocal advanced_services
+            advanced_services = services
+
+        agent_0 = Agent_0()
+        agent_0.advance = local_advance
+        agent_0._log = mock.Mock()
+        agent_0._service_task_ids = list()
 
+        agent_0.services_event = mock.Mock()
+        test_data = [dict({'cores_per_rank': '3', 'executable':'/bin/ls'})]
+        agent_0._cfg = ru.Config(from_dict={'pid':12, 'pilot_sandbox':'/'})
+        agent_0._cfg.services = test_data
+
+        agent_0.services_event.wait = mock.Mock(return_value=True)
+        agent_0._start_services()
+        self.assertTrue(advanced_services[0]['uid'].startswith('service.'))
+        self.assertTrue(test_data[0].items() <=
+                        advanced_services[0]['description'].items())
```

Comment:
the point of this is not clear, better you can just compare known attributes, e.g., `cores_per_rank` and `executable`, and maybe check that some default attributes are set (e.g., `gpus_per_rank=0`, `ranks=1`)

Reply:
The test_data contains `cores_per_rank` and `executable`. As far as I know, `cores_per_rank` is a default attribute.
I am checking both keys and values.

Reply:
@mtitov ? Can you update here